pkg/common/namespace_selection.go

```diff
@@ -5,44 +5,100 @@ package common
+	// responsibility of the calling function for when MatchLabels/MatchExpressions are both nil.
+	matchLabels := map[string]string{}
+	matchExpressions := []metav1.LabelSelectorRequirement{}
+
+	if selector.MatchLabels != nil {
+		matchLabels = *selector.MatchLabels
+	}
+
+	if selector.MatchExpressions != nil {
+		matchExpressions = *selector.MatchExpressions
+	}
+
+	labelSelector = metav1.LabelSelector{
+		MatchLabels:      matchLabels,
+		MatchExpressions: matchExpressions,
+	}
+
+	return labelSelector
+}
+
+// GetSelectedNamespaces returns the list of filtered namespaces according to the policy namespace selector.
+func GetSelectedNamespaces(selector policyv1.Target) ([]string, error) {
+	// Build LabelSelector from provided MatchLabels and MatchExpressions
+	labelSelector := parseToLabelSelector(selector)
 
-	for _, value := range included {
-		found := FindPattern(value, allNamespaces)
-		if found != nil {
-			includedNamespaces = append(includedNamespaces, found...)
-		}
+	// get all namespaces matching selector
+	allNamespaces, err := GetAllNamespaces(labelSelector)
+	if err != nil {
+		log.Error(err, "error retrieving namespaces")
+
+		return []string{}, err
 	}
 
-	// then get the list of excluded
-	excludedNamespaces := []string{}
+	// filter the list based on the included/excluded list of patterns
+	included := selector.Include
+	excluded := selector.Exclude
+	log.V(2).Info("Filtering namespace list using include/exclude lists", "include", included, "exclude", excluded)
+
+	finalList, err := Matches(allNamespaces, included, excluded)
+	if err != nil {
+		return []string{}, err
+	}
 
-	for _, value := range excluded {
-		found := FindPattern(value, allNamespaces)
-		if found != nil {
-			excludedNamespaces = append(excludedNamespaces, found...)
-		}
+	if len(finalList) == 0 {
+		log.V(2).Info("Filtered namespace list is empty.")
 	}
 
-	return DeduplicateItems(includedNamespaces, excludedNamespaces)
+	log.V(2).Info("Returning final filtered namespace list", "namespaces", finalList)
+
+	return finalList, err
 }
 
-// GetAllNamespaces gets the list of all namespaces from k8s
-func GetAllNamespaces() (list []string, err error) {
-	namespaces := KubeClient.CoreV1().Namespaces()
-	namespaceList, err := namespaces.List(context.TODO(), metav1.ListOptions{})
+// GetAllNamespaces gets the list of all namespaces from k8s that matches the input label selector.
+func GetAllNamespaces(labelSelector metav1.LabelSelector) ([]string, error) {
+	parsedSelector, err := metav1.LabelSelectorAsSelector(&labelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing namespace LabelSelector: %w", err)
+	}
+
+	listOpt := metav1.ListOptions{
+		LabelSelector: parsedSelector.String(),
+	}
+
+	log.V(2).Info("Retrieving namespaces with LabelSelector", "LabelSelector", parsedSelector.String())
+
+	nsList, err := (KubeClient).CoreV1().Namespaces().List(context.TODO(), listOpt)
```

Comment:
As I was refactoring to abstract `policyv1.Target`, it started to bug me that these functions were namespace-specific, but that can be addressed if `policyv1.Target` does indeed get used for a different API endpoint since I don't think refactoring it at this point is necessary.